src/dune_rules/pkg_rules.ml

```diff
@@ -1744,40 +2049,43 @@ let setup_rules ~components ~dir ctx =
 ;;
 
 let ocaml_toolchain context =
-  (let* lock_dir = Lock_dir.get context in
-   let* db = DB.get context in
-   match lock_dir.ocaml with
-   | None -> Memo.return `System_provided
-   | Some ocaml -> Resolve.resolve db context ocaml)
-  >>| function
-  | `System_provided -> None
-  | `Inside_lock_dir pkg ->
-    let toolchain =
-      let cookie = (Pkg_installed.of_paths pkg.paths).cookie in
-      let open Action_builder.O in
-      let* cookie = cookie in
-      (* TODO we should use the closure of [pkg] *)
-      let binaries =
-        Section.Map.find cookie.files Bin |> Option.value ~default:[] |> Path.Set.of_list
+  let* lock_dir = Lock_dir.get context in
+  let* db = DB.get context in
+  match lock_dir.ocaml with
+  | None -> Memo.return None
+  | Some ocaml ->
+    let+ pkg = Resolve.resolve db context ocaml in
+    if pkg.system_provided
+    then None
+    else (
+      let toolchain =
+        if Pkg.Toolchain.is_compiler_and_toolchains_enabled pkg
+        then Action_builder.of_memo @@ Pkg.Toolchain.ocaml_toolchain pkg context
+        else (
+          let cookie = (Pkg_installed.of_paths pkg.paths).cookie in
+          let open Action_builder.O in
+          let* cookie = cookie in
+          (* TODO we should use the closure of [pkg] *)
+          let binaries =
+            Section.Map.find cookie.files Bin
+            |> Option.value ~default:[]
+            |> Path.Set.of_list
+          in
+          let env = Env.extend_env (Global.env ()) (Pkg.exported_env pkg) in
+          let path = Env_path.path (Global.env ()) in
+          Action_builder.of_memo @@ Ocaml_toolchain.of_binaries ~path context env binaries)
       in
-      let env = Env.extend_env (Global.env ()) (Pkg.exported_env pkg) in
-      let path = Env_path.path (Global.env ()) in
-      Action_builder.of_memo @@ Ocaml_toolchain.of_binaries ~path context env binaries
-    in
-    Some (Action_builder.memoize "ocaml_toolchain" toolchain)
+      Some (Action_builder.memoize "ocaml_toolchain" toolchain))
 ;;
 
 let all_packages context =
   let* db = DB.get context in
   Dune_lang.Package_name.Map.values db.all
   |> Memo.parallel_map ~f:(fun (package : Lock_dir.Pkg.t) ->
     let package = package.info.name in
-    Resolve.resolve db context (Loc.none, package)
-    >>| function
-    | `Inside_lock_dir pkg -> Some pkg
-    | `System_provided -> None)
-  >>| List.filter_opt
+    Resolve.resolve db context (Loc.none, package))
   >>| Pkg.top_closure
+  >>| List.filter ~f:(Fun.negate Pkg.is_system_provided)
```

Comment:
doesn't seem need to do this filtering in two different places. can we unify this?